### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by dham**
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.

In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.

Click the **Files changed** or **Commits** tab to see all of the changes pushed to `master` since the assignment started. Your teacher can see this too.

<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

Use this PR to leave feedback. Here are some tips:

- Click the **Files changed** tab to see all of the changes pushed to `master` since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to `master`. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.

For more information about this pull request, read “[Leaving assignment feedback in GitHub](https://docs.github.com/education/manage-coursework-with-github-classroom/leave-feedback-with-pull-requests)”.
</details>

Subscribed: @indiamai

### CODE REVIEW

File: fe_utils/quadrature.py [hunk 1]
```diff
@@ -43,8 +43,8 @@ def integrate(self, function):
         The implementation of this method is left as an :ref:`exercise
         <ex-integrate>`.
         """
-
-        raise NotImplementedError
+        # raise "Not implemented error"
```
Comment (dham) [inline fe_utils/quadrature.py:46]: Don't leave your rubbish lying around ;)

**Review by dham — COMMENTED** [review]
As we discussed, this is absolutely fine. It's all correct, you haven't fallen into any algorithmic traps, and I'm not going to try to teach coding style to you!

File: fe_utils/function_spaces.py [hunk 2]
```diff
@@ -23,15 +41,15 @@ def __init__(self, mesh, element):
...
 
         # Implement global numbering in order to produce the global
         # cell node list for this space.
         #: The global cell node list. This is a two-dimensional array in
         #: which each row lists the global nodes incident to the corresponding
         #: cell. The implementation of this member is left as an
         #: :ref:`exercise <ex-function-space>`
-        self.cell_nodes = None
+        self.cell_nodes = compute_cell_nodes(mesh,element)
+        print(self.cell_nodes)
```
Comment (dham) [inline fe_utils/function_spaces.py]: Um WTF?

File: fe_utils/mesh.py [hunk 3]
```diff
@@ -111,7 +112,12 @@ def jacobian(self, c):
         :result: The Jacobian for cell ``c``.
         """
 
-        raise NotImplementedError
+        cg1 = LagrangeElement(self.cell, 1)
+        cg1fs = FunctionSpace(self, cg1)
```
Comment (dham) [inline fe_utils/mesh.py]: If `n` is mesh size then global numbering is `O(n)`, so this makes the cost of evaluating one cell Jacobian `O(n)` and the cost of evaluating an integral over a mesh `O(n^2)`, which is very sad. You need to lift the things that don't depend on `c` up into the mesh constructor.

File: fe_utils/function_spaces.py (hunk 2 above)
Comment (indiamai) [reply fe_utils/function_spaces.py]: whoops, that was a (very useful) debugging statement

File: fe_utils/mesh.py (hunk 3 above)
Comment (indiamai) [reply fe_utils/mesh.py]: makes sense, think i've fixed this now

File: fe_utils/solvers/mastery.py [hunk 4]
```diff
@@ -2,9 +2,133 @@
...
+
+        # # i = gradient component
+        # # ji dimensions of jacobian
+        # J_phi = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
+        # div = np.sum(np.array([J_phi[:, j, :, j] for j in range(d)]), axis=0)
+        # # c = basis of other space
+        # B[np.ix_(nodes2, nodes1)] += np.einsum("q, qc, qb -> cb", Q.weights, phi2, div) * detJ 
+
+        div = np.array([phi1_grad[:, j, :, j] for j in range(d)])
+        Jdiv = np.einsum("ij,jkm -> km", np.linalg.inv(J.T), div)
```
Comment (dham) [inline fe_utils/solvers/mastery.py]: This is definitely wrong. You can't sum out the divergence and then transform. div(u) = du_i/dx_i . Now apply the chain rule to express this in terms of du_i/dX_j  and dx_i/dX_j .

File: fe_utils/solvers/mastery.py [hunk 5]
```diff
@@ -2,9 +2,133 @@
...
+        B[np.ix_(nodes2, nodes1)] += np.einsum("i, ij, ik -> jk", Q.weights, phi2, Jdiv) * detJ 
+
+        # # phi_term = np.einsum("iq, jq -> ijq", phi1.T, phi1.T)
+        
+        # half_grad_term = np.einsum("li, ijkm -> ljkm", np.linalg.inv(J.T), phi1_grad.T)
+        half_grad_term = np.einsum("ji,bdqi -> bdqi", np.linalg.inv(J.T), phi1_grad)
```
Comment (dham) [inline fe_utils/solvers/mastery.py]: This einsum also looks wrong. Which index is the dot product summing over, i or j?

File: fe_utils/solvers/mastery.py [hunk 6]
```diff
@@ -2,9 +2,133 @@
...
+        half_grad_termT = np.einsum("bdqi -> biqd", half_grad_term)
```
Comment (dham) [inline fe_utils/solvers/mastery.py]: This also somehow looks like the wrong transpose, but I find your index numbering hard to follow. If i is the basis function index as in the notes, then this is definitely the wrong transpose.

File: fe_utils/solvers/mastery.py [hunk 7]
```diff
@@ -2,9 +2,133 @@
...
+        # half_grad_termT = np.einsum("ljkm -> kjlm", half_grad_term)
+        grad_term = (1/2)*(half_grad_term + half_grad_termT)
+        
+        # twice_sum = np.einsum("ijkm -> jm", grad_term * grad_term)
+        twice_sum = np.einsum("biqd ->bq ", grad_term * grad_term)
```
Comment (dham) [inline fe_utils/solvers/mastery.py]: I am also not convinced that this is the right einsum for the inner product of grad_term with itself. It's certainly not the obvious way of writing it.

File: fe_utils/solvers/mastery.py (hunk 5 above)
Comment (indiamai) [reply fe_utils/solvers/mastery.py]: yeah this definitely should have j in the rhs instead of i

File: fe_utils/solvers/mastery.py (hunk 4 above)
Comment (indiamai) [reply fe_utils/solvers/mastery.py]: I think i did this right in a comment above- left the wrong bit commented out. Do you transform and then sum out the divergence?

File: fe_utils/solvers/mastery.py (hunk 6 above)
Comment (indiamai) [reply fe_utils/solvers/mastery.py]: i is numbering the derivatives. b is the basis numbering, d is the dimension numbers/components (2), q is the quadrature numbering. I think this is right - it swaps the derivative axis with the components axis.

File: fe_utils/solvers/mastery.py (hunk 7 above)
Comment (indiamai) [reply fe_utils/solvers/mastery.py]: I think I'm rather confused about what is meant to happen here- what exactly does the : in the notes mean?

File: fe_utils/solvers/mastery.py (hunk 6 above)
Comment (dham) [reply fe_utils/solvers/mastery.py]: You're trying to kill me, right? https://finite-element.github.io/5_functions.html#integration
i, j, k... are conventionally basis function indices. q for quadrature, and a, b, g, e (standing in for greek letters) for geometric dimensions.
But yes, swapping derivatives and components is right.

Comment (indiamai) [reply fe_utils/solvers/mastery.py]: at least its somewhat consistent! before i was randomly using ijklm for any axis i wanted. will consider changing it to help your marking!

File: fe_utils/solvers/mastery.py (hunk 7 above)
Comment (dham) [reply fe_utils/solvers/mastery.py]: : is the inner product. So: 
formula

Comment (indiamai) [reply fe_utils/solvers/mastery.py]: alright thank you!

File: fe_utils/solvers/mastery.py (hunk 4 above)
Comment (dham) [reply fe_utils/solvers/mastery.py]: I think that might be right, you're still making it quite complicated. This is completely unchecked but something like:
```
div = np.einsum(`ab,ibqa->iq`, np.linalg.inv(J), phi1_grad)
```
maybe?

Comment (indiamai) [reply fe_utils/solvers/mastery.py]: I think that does what i'm trying to do in two steps in one step - will look into it, thanks!
